packages/service-library/tests/fastapi/test_cancellation_middleware.py

```diff
@@ -0,0 +1,148 @@
+                await server_cancelled_mock()
+            finally:
+                server_done_event.set()
+
+    @router.get("/sleep-with-background-task")
+    async def sleep_with_background_task(
+        sleep_time: float, background_tasks: BackgroundTasks
+    ) -> dict[str, str]:
+        with log_context(logging.INFO, msg="sleeper with background task"):
+            background_tasks.add_task(_sleep_in_the_back, sleep_time)
+            return {"message": "Sleeping in the back"}
+
+    return router
+
+
+@pytest.fixture
+def fastapi_app(fastapi_router: APIRouter) -> FastAPI:
+    app = FastAPI()
+    app.include_router(fastapi_router)
+    app.add_middleware(RequestCancellationMiddleware)
+    return app
+
+
+@pytest.fixture
+def uvicorn_server(fastapi_app: FastAPI) -> Iterator[URL]:
+    random_port = unused_port()
+    with log_context(
+        logging.INFO,
+        msg=f"with uvicorn server on 127.0.0.1:{random_port}",
+    ) as ctx:
+        config = uvicorn.Config(
+            fastapi_app,
+            host="127.0.0.1",
+            port=random_port,
+            log_level="error",
+        )
+        server = uvicorn.Server(config)
+
+        thread = Thread(target=server.run)
+        thread.daemon = True
+        thread.start()
+
+        ctx.logger.info(
+            "server ready at: %s",
+            f"http://127.0.0.1:{random_port}",
+        )
+
+        yield URL(f"http://127.0.0.1:{random_port}")
+
+        server.should_exit = True
+        thread.join(timeout=10)
+
+
+async def test_server_cancels_when_client_disconnects(
+    uvicorn_server: URL,
+    server_done_event: asyncio.Event,
+    server_cancelled_mock: AsyncMock,
+):
+    async with httpx.AsyncClient(base_url=f"{uvicorn_server}") as client:
+        # check standard call still complete as expected
+        with log_context(logging.INFO, msg="client calling endpoint"):
+            response = await client.get("/sleep", params={"sleep_time": 0.1})
+        assert response.status_code == 200
+        assert response.json() == {"message": "Slept for 0.1 seconds"}
+        async with asyncio.timeout(10):
+            await server_done_event.wait()
+        server_done_event.clear()
+
+        # check slow call get cancelled
+        with log_context(
+            logging.INFO, msg="client calling endpoint for cancellation"
+        ) as ctx:
+            with pytest.raises(httpx.ReadTimeout):
+                response = await client.get(
+                    "/sleep", params={"sleep_time": 10}, timeout=0.1
+                )
+            ctx.logger.info("client disconnected from server")
+
+        async with asyncio.timeout(5):
+            await server_done_event.wait()
+        server_cancelled_mock.assert_called_once()
+        server_cancelled_mock.reset_mock()
+        server_done_event.clear()
+
+        # NOTE: shows that FastAPI BackgroundTasks get cancelled too!
+        # check background tasks get cancelled as well sadly
```

Comment:
I'd say it's a good thing, makes us move away form them
```suggestion
        # check background tasks get cancelled as well fortunately
```